### PR TITLE
Make Array.Reverse faster

### PR DESCRIPTION
**Pull request by jkotas**
Array.Reverse showed up as hot method in profiles of real workloads (e.g. it is called from Json.NET frequently via List.Reverse)

### CODE REVIEW

**Comment by jkotas** [comment]
@ahsonkhan This implements the improvement we have discussed over email.

File: src/mscorlib/src/System/Array.cs [hunk 1]
```diff
@@ -1707,13 +1707,14 @@ public static void Reverse<T>(T[] array, int index, int length)
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_InvalidOffLen);
             Contract.EndContractBlock();
 
+            ref T p = ref Unsafe.As<byte, T>(ref array.GetRawSzArrayData());
```
Comment (ahsonkhan) [inline src/mscorlib/src/System/Array.cs]: Is there a way to do something similar to replace TrySZReverse in the non-generic Reverse?
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Array.cs#L1656

File: src/mscorlib/src/System/Array.cs [hunk 2]
```diff
@@ -1707,13 +1707,14 @@ public static void Reverse<T>(T[] array, int index, int length)
...
             int i = index;
             int j = index + length - 1;
             while (i < j)
             {
-                T temp = array[i];
-                array[i] = array[j];
-                array[j] = temp;
+                T temp = Unsafe.Add(ref p, i);
+                Unsafe.Add(ref p, i) = Unsafe.Add(ref p, j);
+                Unsafe.Add(ref p, j) = temp;
```
Comment (stephentoub) [inline src/mscorlib/src/System/Array.cs]: @jkotas, for my edification, what's the explanation for why this is faster?  It's avoiding bounds checks that the JIT is currently unable to elide?

Comment (jkotas) [reply src/mscorlib/src/System/Array.cs]: It is avoiding bounds checks and covariance checks (for arrays of object references).

File: src/mscorlib/src/System/Array.cs (hunk 1 above)
Comment (jkotas) [reply src/mscorlib/src/System/Array.cs]: Replacing TrySZReverse with C# implementation would be more profound change. I will leave it for a different day..

File: src/mscorlib/src/System/Array.cs (hunk 2 above)
Comment (stephentoub) [reply src/mscorlib/src/System/Array.cs]: Got it.

File: src/mscorlib/src/System/Array.cs [hunk 3]
```diff
@@ -1657,22 +1657,15 @@ public static void Reverse(Array array, int index, int length)
...
             {
-                while (i < j)
-                {
-                    Object temp = objArray[i];
-                    objArray[i] = objArray[j];
-                    objArray[j] = temp;
-                    i++;
-                    j--;
-                }
+                Array.Reverse<object>(objArray, index, length);
```
Comment (ahsonkhan) [inline src/mscorlib/src/System/Array.cs]: Nice.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 corefx_baseline

File: src/mscorlib/src/System/Array.cs (hunk 2 above)
Comment (benaadams) [reply src/mscorlib/src/System/Array.cs]: `List<T>` is non-covariant so this could be done for `Add` etc?

Comment (jkotas) [reply src/mscorlib/src/System/Array.cs]: For List, you need to be careful about corrupted state that can lead to buffer overruns or type safety violations. Some of these bugs may be in the user code, but our general standard is that user code race should not turn into type or memory safety violations. I think a better way to deal with the covariance check overhead in general case like List is something like the covariance array wrapper (https://github.com/dotnet/corefx/issues/23689) - the unsafe code is just one line in the covariance array wrapper. The rest is safe.

Comment (ViktorHofer) [reply src/mscorlib/src/System/Array.cs]: > In theory, the JIT can notice that the source of the value is the same array as what it is assigned to and use that information to ship the covariance check ...

Is skipping the covariance when source and destination are same tracked somewhere?

Comment (benaadams) [reply src/mscorlib/src/System/Array.cs]: https://github.com/dotnet/coreclr/issues/14574

Comment (AndyAyersMS) [reply src/mscorlib/src/System/Array.cs]: If you load and store to the same array directly the jit already does this optimization (see for instance #14580, which enhanced an existing optimization). But if load and store are disconnected the jit may miss out.

@erozenfeld maybe this is something we could do more generally, once we have built up dataflow?

**Pull request merged**